src/api/docroot.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+// DO NOT SUBMIT: is DocRoot concept enough? Or do we also need a separate
+// concept of Scope?
+
+/**
+ * `DocRoot` is very similar to the `Document` interface and acts as a stand-in
+ * for the `Document` in most of cases. It's a pseudo-document and a unit of
+ * scoping in AMP. Each independent sub-document will have its own instance of
+ * `DocRoot` even if multiple `DocRoot`s share the same global `Document`
+ * instance.
+ *
+ * @interface
+ */
+export class DocRoot {
+
+  /**
+   * Returns the current URL of the docroot.
+   * See https://developer.mozilla.org/en-US/docs/Web/API/Document/URL.
+   *
+   * @return {string}
+   */
+  get url() {}
```

Comment:
This is a special case but it would be great to keep in mind "root"-like use cases. 

One example is service scoping for things other than AMP documents, e.g. amp-script having its own URL replacement service for its subtree. I have a draft PR that adapts `AmpDocShadow` for this: https://github.com/ampproject/amphtml/pull/22327/files#diff-b596009fc4aa42dfa7d85c56dccf78eeR489

Reply:
I'm also very interested in root-like use cases. Can we discuss in a greater detail? At the very top, I have a note:

> // DO NOT SUBMIT: is DocRoot concept enough? Or do we also need a separate
> // concept of Scope?

Root-like cases is what I meant there. I've been laboring a bit with the concept of `documentElement` vs `head` vs `body`, that are not as easy for other types of docs to map. Even shadow docs have issues here. However, I also wanted to make sure that any extension can be installed inside a root and as it stands a lot of extensions have very document-like concepts and expect document structure. That's why I called it "docroot" and not just "root" or a "scope". So, I'm very interested in flashing this out more for root.